### PR TITLE
fix(cb2-5499): update test station details

### PR DESCRIPTION
**Pull request by gjulien-bjss**
## Test station type does not update on selection

_Test station type does not update on selection_
[CB2-5499](https://dvsa.atlassian.net/browse/CB2-5499)

## Checklist

- [ ] Branch is rebased against the latest develop/common
- [ ] Necessary `id` required prepended with `"test-"` have been checked with automation testers and added
- [ ] Code and UI has been tested manually after the additional changes
- [ ] PR title includes the JIRA ticket number
- [ ] Squashed commits contain the JIRA ticket number
- [ ] Link to the PR added to the repo
- [ ] Delete branch after merge

### CODE REVIEW

**Review by Ruairidh-BJSS — DISMISSED** [review]

**Review by CormacBJSS — APPROVED** [review]

File: src/app/store/test-records/reducers/test-records.reducer.ts [hunk 1]
```diff
@@ -65,7 +66,20 @@ export const testResultsReducer = createReducer(
...
+    return {
+      ...state,
+      editingTestResult: {
+        ...state.editingTestResult,
+        testStationName: action.payload.testStationName,
+        testStationType: action.payload.testStationType
+      }
+    };
+  })
 );
```
Comment (tahayazici) [inline src/app/store/test-records/reducers/test-records.reducer.ts:83]: Just to make it a bit more concise
```suggestion
  on(updateTestStation, (state, action) => !state.editingTestResult
    ?  { ...state }
    : {
      ...state,
      editingTestResult: {
        ...state.editingTestResult,
        testStationName: action.payload.testStationName,
        testStationType: action.payload.testStationType
      }
    }
);
```

File: src/app/store/test-stations/actions/test-stations.actions.ts [hunk 2]
```diff
@@ -15,3 +15,5 @@ function getTitle(isPlural: boolean = false, suffix: string = ''): string {
   suffix = suffix ? ' ' + suffix : suffix;
   return '[API/test-stations] Fetch Test Station' + plural + suffix;
 }
+
+export const updateTestStation = createAction('[test-stations] update the test station', props<{ payload: TestStation }>());
```
Comment (tahayazici) [inline src/app/store/test-stations/actions/test-stations.actions.ts:19]: You could extend `getTitle()` to take in an `action` parameter and use that.

File: src/app/forms/validators/custom-async-validators.ts [hunk 3]
```diff
@@ -24,4 +28,15 @@ export class CustomAsyncValidators {
...
       );
   }
+
+  static updateTestStationDetails(store: Store<State>): AsyncValidatorFn {
+    return (control: AbstractControl): Observable<null> => {
+      store
+        .pipe(select(getTestStationFromProperty((control as CustomFormControl).meta.name as keyof TestStation, control.value)), take(1))
+        .subscribe(stations => {
+          stations && store.dispatch(updateTestStation({ payload: stations }));
+        });
```
Comment (tahayazici) [inline src/app/forms/validators/custom-async-validators.ts:38]: You were thinking of where to use `tap()`, right? 🙂
```suggestion
      const propertyName = (control as CustomFormControl).meta.name as keyof TestStation;
      
      store
        .pipe(
          select(getTestStationFromProperty(propertyName, control.value)),
          take(1),
          tap(stations => stations && store.dispatch(updateTestStation({ payload: stations })))
        )
        .subscribe();
```

File: src/app/store/test-records/reducers/test-records.reducer.ts (hunk 1 above)
Comment (gjulien-bjss) [reply src/app/store/test-records/reducers/test-records.reducer.ts:83]: Nice

File: src/app/forms/validators/custom-async-validators.ts (hunk 3 above)
Comment (gjulien-bjss) [reply src/app/forms/validators/custom-async-validators.ts:38]: I did think about using the tap, but decided against it and just put it in the subscribe instead

File: src/app/store/test-stations/actions/test-stations.actions.ts (hunk 2 above)
Comment (tahayazici) [reply src/app/store/test-stations/actions/test-stations.actions.ts:19]: That's fair.

File: src/app/forms/validators/custom-async-validators.ts (hunk 3 above)
Comment (tahayazici) [reply src/app/forms/validators/custom-async-validators.ts:38]: Don't you think it looks cleaner within `tap()`? It's like an effect of the store - do go and trigger something else after getting the value of the observable.

**Review by kpearmanbjss — APPROVED** [review]

File: src/app/forms/validators/custom-async-validators.ts (hunk 3 above)
Comment (mcadenas-bjss) [reply src/app/forms/validators/custom-async-validators.ts:38]: feels like that would be using an operator for the sake of it. You have to subscribe anyway so might as well use it.

**Review by mcadenas-bjss — APPROVED** [review]

**Pull request merged**